src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/BulkCopy/SqlBulkCopy.Async.cs

```diff
@@ -0,0 +1,1108 @@
+                }
+            }
+            return resultTask;
+        }
+
+        private async Task<bool> ReadFromRowSourceAsync(CancellationToken cts)
+        {
+            if (_DbDataReaderRowSource != null)
+            {
+                // This will call ReadAsync for DbDataReader (for SqlDataReader it will be truly async read; for non-SqlDataReader it may block.)
+                return await _DbDataReaderRowSource.ReadAsync(cts).ContinueWith((t) =>
+                {
+                    if (t.Status == TaskStatus.RanToCompletion)
+                    {
+                        _hasMoreRowToCopy = t.Result;
+                    }
+                    return t;
+                }, TaskScheduler.Default).Unwrap();
+            }
+            else
+            { // This will call Read for DataRows, DataTable and IDataReader (this includes all IDataReader except DbDataReader)
+              // Release lock to prevent possible deadlocks
+                SqlInternalConnectionTds internalConnection = _connection.GetOpenTdsConnection();
+                bool semaphoreLock = internalConnection._parserLock.CanBeReleasedFromAnyThread;
+                internalConnection._parserLock.Release();
+
+                _hasMoreRowToCopy = false;
+                try
+                {
+                    _hasMoreRowToCopy = ReadFromRowSource(); // Synchronous calls for DataRows and DataTable won't block. For IDataReader, it may block.
+                }
+                catch (Exception ex)
+                {
+                    return await Task.FromException<bool>(ex);
+                }
+                finally
+                {
+                    internalConnection._parserLock.Wait(canReleaseFromAnyThread: semaphoreLock);
+                }
+                return false;
+            }
+        }
+
+        // Rest of the WriteToServerInternalAsync method.
+        // It carries on the source from its caller WriteToServerInternal.
+        // source is null in case of Sync bcp. But valid in case of Async bcp.
+        // It calls the WriteToServerInternalRestContinuedAsync as a continuation of the initial query task.
+        private void WriteToServerInternalRestAsync(CancellationToken cts, TaskCompletionSource<object> source)
+        {
+            Debug.Assert(_hasMoreRowToCopy, "first time it is true, otherwise this method would not have been called.");
+            _hasMoreRowToCopy = true;
+            Task<BulkCopySimpleResultSet> internalResultsTask = null;
+            BulkCopySimpleResultSet internalResults = new BulkCopySimpleResultSet();
+            SqlInternalConnectionTds internalConnection = _connection.GetOpenTdsConnection();
+            try
+            {
+                _parser = _connection.Parser;
+                _parser._asyncWrite = true; // Very important!
+
+                Task reconnectTask;
+                try
+                {
+                    reconnectTask = _connection.ValidateAndReconnect(
```

Comment:
async capture, is it possible to pass the parserLock as a state parameter?

Reply:
I have not spend much time in `reconnect task` in `SqlConnection`. It follows the same steps, if the task returns null/task. I am not sure if `AsyncHelper` could be called to pass state on it. I will look into that tomorrow. In my opinion AsyncHelper is one the parts that we need to remove or at least simplify it with `Task`'s built-in `capabilities`.

Reply:
Ideally AsyncHelper wouldn't be needed at all, i agree. Since it is currently needed and I've spent time around it this one stuck out to me while i was reviewing the code. If you'd prefer not to change this then I'm ok with that.